jobrunner/run.py

```diff
@@ -195,28 +161,161 @@ def handle_running_job(job):
-        raise
+    # handle the simple no change needed states.
+    if old_status.state in STABLE_STATES:
+        # no action needed, simply update job message and timestamp
+        message = old_status.state.value.title()
+        set_message(job, message)
+        return
+
+    # ok, handle the state transitions that are our responsibility
+    if old_status.state == ExecutorState.UNKNOWN:
+        # a new job
+        if job.state == State.RUNNING:
+            log.warning("Got UNKNOWN state for a job we thought was RUNNING")
+
+        # check dependencies
+        awaited_states = get_states_of_awaited_jobs(job)
+        if State.FAILED in awaited_states:
+            mark_job_as_failed(
+                job,
+                "Not starting as dependency failed",
+                code=StatusCode.DEPENDENCY_FAILED,
+            )
+            return
+
+        if any(state != State.SUCCEEDED for state in awaited_states):
+            set_message(
+                job, "Waiting on dependencies", code=StatusCode.WAITING_ON_DEPENDENCIES
+            )
+            return
+
+        next_state = ExecutorState.PREPARING
+        new_status = api.prepare(definition)
+
+    elif old_status.state == ExecutorState.PREPARED:
+        next_state = ExecutorState.EXECUTING
+        new_status = api.execute(definition)
+
+    elif old_status.state == ExecutorState.EXECUTED:
+        next_state = ExecutorState.FINALIZING
+        new_status = api.finalize(definition)
+
+    elif old_status.state == ExecutorState.FINALIZED:
+        # final state - we have finished!
+        results = api.get_results(definition)
+        # TODO: implement workspace API
+        # delete_obsolete_files(job, results)
+        save_results(job, results)
+        mark_job_as_completed(job)
+        api.cleanup(definition)
+        # we are done here
+        return
+
+    # following logic is common to all non-final transitions
+
+    if new_status.state == old_status.state:
+        # no change in state, i.e. back pressure
+        set_message(
+            job, "Waiting on available resources", code=StatusCode.WAITING_ON_WORKERS
+        )
+
+    elif new_status.state == next_state:
+        # successful state change to the expected next state
+        if new_status.state == ExecutorState.PREPARING:
+            job.state = State.RUNNING
+        elif job.state != State.RUNNING:
+            # got an ExecutorState that should mean the job.state is RUNNING, but it is not
+            log.warning("Got {new_status.state} for job we thought was {job.state}")
+        set_message(job, new_status.state.value.title())
+
+    elif new_status.state == ExecutorState.ERROR:
+        # all transitions can go straight to error
+        mark_job_as_failed(job, new_status.message)
+        api.cleanup(definition)
+
+    else:
+        raise InvalidTransition(
+            f"unexpected state transition from {old_status.state} to {new_status.state}: {new_status.message}"
+        )
```

Comment:
Do we need to include the job details here, or is it always guaranteed to be logged by whatever handles this exception?

Reply:
As long as the exception is logged withing the with context manager, it should have an id. I'll add it explicitly in case it bubbles up beyond for some reason.